Kernel/Devices/GPU/3dfx/VoodooDisplayConnector.cpp

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright (c) 2023, Edwin Rijkee <edwin@virtualparadise.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Devices/GPU/3dfx/VoodooDisplayConnector.h>
+#include <Kernel/Devices/GPU/Console/ContiguousFramebufferConsole.h>
+#include <Kernel/Devices/GPU/Management.h>
+
+namespace Kernel::VoodooGraphics {
+
+NonnullLockRefPtr<VoodooDisplayConnector> VoodooDisplayConnector::must_create(PhysicalAddress framebuffer_address, size_t framebuffer_resource_size, Memory::TypedMapping<RegisterMap volatile> registers_mapping, NonnullOwnPtr<IOWindow> io_window)
+{
+    auto device_or_error = DeviceManagement::try_create_device<VoodooDisplayConnector>(framebuffer_address, framebuffer_resource_size, move(registers_mapping), move(io_window));
+    VERIFY(!device_or_error.is_error());
+    auto connector = device_or_error.release_value();
+    MUST(connector->create_attached_framebuffer_console());
+    MUST(connector->fetch_and_initialize_edid());
+    return connector;
+}
+
+ErrorOr<void> VoodooDisplayConnector::fetch_and_initialize_edid()
+{
+    // TODO: actually fetch the EDID.
+    return initialize_edid_for_generic_monitor({});
+}
+
+ErrorOr<void> VoodooDisplayConnector::create_attached_framebuffer_console()
+{
+    m_framebuffer_console = Graphics::ContiguousFramebufferConsole::initialize(m_framebuffer_address.value(), 1024, 768, 1024 * sizeof(u32));
+    GraphicsManagement::the().set_console(*m_framebuffer_console);
+    return {};
+}
+
+VoodooDisplayConnector::VoodooDisplayConnector(PhysicalAddress framebuffer_address, size_t framebuffer_resource_size, Memory::TypedMapping<RegisterMap volatile> registers_mapping, NonnullOwnPtr<IOWindow> io_window)
+    : DisplayConnector(framebuffer_address, framebuffer_resource_size, true)
+    , m_registers(move(registers_mapping))
+    , m_io_window(move(io_window))
+{
+}
+
+// clang-format off
+static constexpr DisplayConnector::ModeSetting known_modes[] = {
```

Comment:
Are these modes hardcoded because this is what the card supports? what about getting these modes from the EDID?

Reply:
No, the card is more flexible, but when `set_mode_setting` is called the mode never seems to have any of the detailed timing fields set. It only receives pixels width and height. I looked at getting it from the EDID but only the 'detailed timing' have all the necessary values, there is often only one detailed timing: the native resolution of the monitor. This driver is still using a dummy EDID created by `initialize_edid_for_generic_monitor()` so it will probably not have the native resolution of the monitor at all.

For the 'established timings' and 'standard timings' in the EDID the only part of this you will get is width+height+refresh rate. The other timing values you either need to know them or you need to calculate them. I didn't see any functions in SerenityOS that do this yet and it's not a card specific thing, so it would probably belong in a shared location.

If `set_mode_setting` is called with all the information included then the driver will use it and ignore the `known_modes`  list. DisplaySettings also still ignores the EDID for resolutions and has a hardcoded list (it uses it for the monitor names and dimensions though).

Reply:
I just found DMT.cpp which seems to have lists of timing data, I will take another look if this can be used.

Reply:
I have added code to select from the modes specified in the EDID (detailed modes or standard/established modes for which there is a DMT mode) and edited the generic EDID.

Perhaps this code should be moved to DisplayConnector, since it seems like something that would apply in other drivers too. Or maybe in the future `set_mode_setting` should always be given a fully specified mode to set :)

Reply:
> Perhaps this code should be moved to DisplayConnector, since it seems like something that would apply in other drivers too. Or maybe in the future `set_mode_setting` should always be given a fully specified mode to set :)

Yeah, the future GPU subsystem will be more tied to bare metal hardware and as such will use the detailed timings more extensively. For now, this is how it works, so as long as you use `LibEDID` code here to get the timings I think it's fine.